pkg/chunk/composite_store.go

```diff
@@ -34,84 +33,140 @@ func (a byStart) Len() int           { return len(a) }
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V4SchemaFrom.Time,
+			SchemaFunc: v4Schema,
+		})
 	}
 
 	if schemaCfg.V5SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v5Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V5SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V5SchemaFrom.Time,
+			SchemaFunc: v5Schema,
+		})
 	}
 
 	if schemaCfg.V6SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v6Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V6SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V6SchemaFrom.Time,
+			SchemaFunc: v6Schema,
+		})
 	}
 
 	if schemaCfg.V7SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v7Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V7SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V7SchemaFrom.Time,
+			SchemaFunc: v7Schema,
+		})
 	}
 
 	if schemaCfg.V8SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v8Schema(schemaCfg), storage)
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V8SchemaFrom.Time,
+			SchemaFunc: v8Schema,
+		})
+	}
+
+	if schemaCfg.V9SchemaFrom.IsSet() {
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V9SchemaFrom.Time,
+			SchemaFunc: v9Schema,
+		})
+	}
+
+	return opts
+}
+
+// StorageOpt stores when a StorageClient is to be used.
+type StorageOpt struct {
+	From   model.Time
+	Client StorageClient
+}
+
+// NewStore creates a new Store which delegates to different stores depending
+// on time.
+func NewStore(cfg StoreConfig, schemaCfg SchemaConfig, schemaOpts []SchemaOpt, storageOpts []StorageOpt) (Store, error) {
+	stores := []compositeStoreEntry{}
+	from := schemaOpts[0].From
+
+	i, j := 0, 0
+
+	for i+1 < len(schemaOpts) && j+1 < len(storageOpts) {
+		currSchema := schemaOpts[i]
+		currStorage := storageOpts[j]
+
+		nextSchema := schemaOpts[i+1]
+		nextStorage := storageOpts[j+1]
+
+		store, err := newStore(cfg, currSchema.SchemaFunc(schemaCfg), currStorage.Client)
 		if err != nil {
 			return nil, err
 		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V8SchemaFrom.Time, store})
+
+		if nextSchema.From.Before(nextStorage.From) {
+			from = nextSchema.From
+			i++
+		} else if nextSchema.From.After(nextStorage.From) {
+			from = nextStorage.From
+			j++
+		} else {
+			from = nextSchema.From
+			i++
+			j++
+		}
+
+		stores = append(stores, compositeStoreEntry{from, store})
 	}
 
-	if schemaCfg.V9SchemaFrom.IsSet() {
-		store, err := newSeriesStore(cfg, v9Schema(schemaCfg), storage)
+	// Now cover the remaining schemas and storages.
+	for i < len(schemaOpts) && j < len(storageOpts) {
+		store, err := newStore(cfg, schemaOpts[i].SchemaFunc(schemaCfg), storageOpts[j].Client)
```

Comment:
You're always calling `newStore`; v9 schema should be used with `newSeriesStore`.

Reply:
Fixed this. Sorry didn't notice.